docs/user-docs/infrastructure_pipeline.md

```diff
@@ -0,0 +1,146 @@
+2. After workspace clean-up on clean code (before artifact publishing)
+
+Common use cases include:
+
+- File transformations or substitutions
+- Environment-specific configuration setup
+- Secret retrieval and file generation
+- Custom validation or security scanning
+
+```yaml
+TerraformBuildInjectionSteps:
+  - task: PowerShell@2
+    displayName: 'Generate backend config'
+    inputs:
+      targetType: 'inline'
+      script: |
+        # Generate backend.tf from environment variables
+        Write-Host "Generating backend configuration..."
+
+  - task: AzureCLI@2
+    displayName: 'Retrieve secrets'
+    inputs:
+      azureSubscription: 'service-connection'
+      scriptType: 'bash'
+      scriptLocation: 'inlineScript'
+      inlineScript: |
+        az keyvault secret show --vault-name myvault --name terraform-vars
+```
+
+### Advanced Usage
+
+Listed below are possible advanced usages.
+
+_If you have any advanced usages, please consider contributing them to the documentation._
+
+#### Custom Agent Pool
+
+```yaml
+extends:
+  template: pipelines/infrastructure_pipeline.yml@templates
+  parameters:
+    PipelinePool: 'MyCustomPool'
+    RelativePathToTerraformFiles: 'infrastructure/terraform'
+    TerraformVersion: '1.6.0'
+```
+
+#### Injection Step to add required_version
+
+```yaml
+extends:
+  template: pipelines/infrastructure_pipeline.yml@PipelineTemplates
+  parameters:
+    RelativePathToTerraformFiles: tests/pipelines/infrastructure_pipeline/terraform
+    PipelinePool: "Mare Nubium"
+    TerraformVersion: 1.1.9
+    TerraformBuildInjectionSteps:
+      - pwsh: |
+          $path = "$(Pipeline.Workspace)/$(Build.Repository.Name)/tests/pipelines/infrastructure_pipeline/terraform/main.tf"
+          $content = Get-Content $path
+          $terraformStart = $content.IndexOf($($content | Where-Object { $_ -match "^terraform\s*{" }))
+          if ($terraformStart -ge 0) {
+            $insertIndex = $terraformStart + 1
+            $content = $content[0..($insertIndex-1)] + '  required_version = "1.1.9"' + $content[$insertIndex..($content.Count-1)]
+            Set-Content $path $content
+          }
+        displayName: "Injecting into terraform block 'required_version'"
+```
+
+#### Pipeline Flow
+
+1. **Repository Checkout**: Checks out the source repository to the build agent
+2. **Injection Steps (First Run)**: Executes any custom steps provided via `TerraformBuildInjectionSteps`
+3. **Terraform Installation**: Installs the specified version of Terraform CLI
+4. **Terraform Init**: Initialises Terraform with backend disabled (`-backend=false`)
+5. **Terraform Validate**: Validates the Terraform configuration files
+6. **Workspace Clean-up**: Performs a clean checkout to ensure artifact purity
+7. **Injection Steps (Second Run)**: Re-executes injection steps on clean code
+8. **Artifact Publishing**: Publishes the Terraform files as a pipeline artifact named "TerraformArtifact"
+
+### Notes
```

Comment:
Consider inline notes with the `> [!NOTE]` block to increase visibility at the right context.